### PR TITLE
Enhance product analysis and comparison

### PR DESCRIPTION
**Pull request by mfarooq2**
This commit enhances the product analysis and comparison capabilities of the application. The changes include:

- Merged essential elements from  into  to consolidate core logic.
- Modified  to extract product specs, pros, and cons.
- Improved the  function to provide a detailed comparison and identify the best product.
- Modified the  function in  to use the  and .
- Updated  to display the enhanced product information.
- Updated unit tests in  and  to cover the merged functionality and new features.
- Added tests for edge cases.

These changes improve the accuracy and depth of the product analysis and comparison.

## Summary by Sourcery

This pull request enhances the product analysis and comparison capabilities of the application by introducing new tools and agents for searching and comparing products across multiple platforms. It also improves the extraction of product information and updates the display of product details.

New Features:
- Introduces a new ProductSearchAgent class that orchestrates product searches and comparisons using the ProductSearchTool and ProductComparisonTool.
- Adds a ProductComparisonTool class to compare product search results using a specified model.
- Adds a ProductSearchTool class to search for products across multiple platforms (Amazon, Ebay, BestBuy).

Enhancements:
- Refactors the product search functionality to use a ProductSearchAgent for orchestrating searches and comparisons.
- Improves product data extraction to include product specifications, pros, and cons.
- Enhances the product comparison function to provide a more detailed comparison and identify the best product based on available information.
- Updates the application to display enhanced product information, including specs, pros, and cons, in a structured table format.

Tests:
- Adds new unit tests to cover the merged functionality and new features, including tests for edge cases and individual components like Amazon, Ebay, and BestBuy search functionalities.
- Adds tests for edge cases.

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -9,6 +9,39 @@
...
+from typing import List, Optional
+from pydantic import BaseModel, Field
+from urllib.parse import quote_plus, urljoin
+import json
+import re
+from transformers import pipeline
+from flask import Flask, render_template
+
+
+# Configure logging
```
[the remark this answers is not in the document]
Comment (mfarooq2) [reply app.py:28]: @sourcery-ai issue

File: app.py [hunk 2]
```diff
@@ -188,15 +316,17 @@ def compare_products_with_ai(products: List[Product]) -> str:
...
 def index():
-    # Example: Searching for "laptop"
     query = "laptop"
-    search_tool = ProductSearchTool()
+    agent = ProductSearchAgent(model_name="google/flan-t5-xxl")
     loop = asyncio.new_event_loop()
     asyncio.set_event_loop(loop)
-    products = loop.run_until_complete(search_tool.search_products(query))
+    
+    search_results = loop.run_until_complete(agent.search_tool.search_products(query))
```
[the remark this answers is not in the document]
Comment (mfarooq2) [reply app.py:324]: @sourcery-ai issue

File: app.py [hunk 3]
```diff
@@ -165,21 +216,98 @@ async def search_products(self, query: str) -> List[Product]:
...
+        self.comparison_tool = ProductComparisonTool()
+        self.model_name = model_name
+
+    async def search_and_compare(self, query: str) -> str:
+        search_results = await self.search_tool.search_products(query)
+        return self.comparison_tool.compare_products(search_results)
+
+
+# Use HuggingFace Transformers for product comparison
+def compare_products_detailed(products: List[Product]) -> str:
```
[the remark this answers is not in the document]
Comment (mfarooq2) [reply app.py:284]: @sourcery-ai issue

File: test_agent.py [hunk 4]
```diff
@@ -49,6 +49,164 @@ def test_parse_search_results(self):
...
+    def test_parse_search_results(self):
+        html = "<html><body><div data-component-type='s-search-result'><h2 class='a-size-medium a-color-base s-inline s-access-title a-text-normal'><a><span>Test Title</span></a></h2><span class='a-price-whole'>$100.00</span></div></body></html>"
+        expected_results = [{"title": "Test Title", "price": 100.00, "url": "https://www.bestbuy.com/test"}]
+        with patch.object(BeautifulSoup, "BeautifulSoup", return_value=BeautifulSoup(html, "html.parser")):
+            self.assertEqual(self.bestbuy.parse_search_results(html), expected_results)
+        
+
+class TestProductSearchTool(unittest.TestCase):
+    def setUp(self):
+        self.product_search_tool = ProductSearchTool()
```
[the remark this answers is not in the document]
Comment (mfarooq2) [reply test_agent.py:129]: @sourcery-ai issue

File: test_product_comparison.py [hunk 5]
```diff
@@ -19,27 +19,43 @@ def test_search_and_compare(self):
...
+        expected_response = "Best product: Product(title='Test Title', price=100.0, url='https://www.amazon.com/test', specs='Specs', pros=['Pro1', 'Pro2'], cons=['Cons1', 'Cons2'])\nComparison: Product: Test Title\nSpecs: Specs\nPros: Pro1, Pro2\nCons: Cons1, Cons2\nProduct: Test Title\nSpecs: Specs\nPros: Pro1, Pro2\nCons: Cons1, Cons2"
+        response = compare_products_detailed(products)
+        self.assertEqual(response, expected_response)
 
 class TestProductSearchTool(unittest.TestCase):
-    @patch('agent1.ProductSearchTool.__init__')
     def setUp(self):
-        self.search_tool = ProductSearchTool()
-        self.query = "test query"
+        self.product_search_tool = ProductSearchTool()
```
[the remark this answers is not in the document]
Comment (mfarooq2) [reply test_product_comparison.py:45]: @sourcery-ai issue

File: test_agent.py [hunk 6]
```diff
@@ -49,6 +49,164 @@ def test_parse_search_results(self):
...
+        expected_results = [
+            {"title": "Test Title", "price": 100.00, "url": "https://www.amazon.com/test", "specs": "Specs", "pros": ["Pro1", "Pro2"], "cons": ["Cons1", "Cons2"]},
+            {"title": "Test Title", "price": 100.00, "url": "https://www.ebay.com/test", "specs": "Specs", "pros": ["Pro1", "Pro2"], "cons": ["Cons1", "Cons2"]},
+            {"title": "Test Title", "price": 100.00, "url": "https://www.bestbuy.com/test", "specs": "Specs", "pros": ["Pro1", "Pro2"], "cons": ["Cons1", "Cons2"]}
+        ]
+        for platform, expected in zip(platforms, expected_results):
+            with patch.object(platform, "make_request", return_value="dummy raw response"):
+                with patch.object(platform, "parse_search_results", return_value=expected):
+                    result = self.product_search_tool._search_platform(platform, query, session=self.session)
+                    self.assertEqual(result, expected)
```
[the remark this answers is not in the document]
Comment (mfarooq2) [reply test_agent.py:143]: @sourcery-ai issue

File: app.py [hunk 7]
```diff
@@ -165,21 +216,98 @@ async def search_products(self, query: str) -> List[Product]:
...
+            comparison_text += f"Cons: {', '.join(product.cons)}\n"
+    
+    
+    best_product = None
+
+    if len(products) > 0:
+        best_product = products[0]
+        best_product_title = products[0].title
+    else:
+        best_product = "No products to compare"
```
[the remark this answers is not in the document]
Comment (mfarooq2) [reply app.py:306]: @sourcery-ai issue
